src/Service/Database/Shell.php

```diff
@@ -0,0 +1,318 @@
+        $compressedFinalFile = $fileNamePrefix . '.sql.gz';
+
+        $databaseName = $this->config->getDatabaseCredentials()->getName();
+
+        /** @var Process[] $commands */
+        $commands = [];
+
+        if (trim($strip_tables) !== '') {
+            // Create the structure-only dump for tables that we don't want the data from.
+            $commands[] = $this->createDumpProcess([
+                '--no-data',
+                '--add-drop-table',
+                // @TODO Do we need to add --skip-triggers here?
+                // @TODO Do we need to add --routines here?
+                $databaseName,
+                $strip_tables,
+                "> $structureOutputFile"
+            ]);
+        } else {
+            // Empty structure file to make the rest of the code more consistent.
+            file_put_contents($structureOutputFile, '');
+        }
+
+        $dataDumpOptions = array_map(function ($table) use ($databaseName) {
+            return sprintf('--ignore-table=%s', escapeshellarg($databaseName . '.' . $table));
+        }, explode(' ', $strip_tables));
+
+        $commands[] = $this->createDumpProcess(array_merge($dataDumpOptions, [
+            '--add-drop-table',
+            // @TODO Do we need to add --skip-triggers here?
+            $databaseName,
+            "> $dataOutputFile"
+        ]));
+
+        $dumpHeader = sprintf(
+            '-- Generator: %s (%s) at %s on %s by %s\n--',
+            Application::APP_NAME,
+            Application::APP_VERSION,
+            date('c'),
+            gethostname(),
+            get_current_user()
+        );
+
+        $this->logger->info('Starting structure and data dump commands.');
+
+        // Kick the commands off
+        array_walk($commands, function (Process $command) {
+            $command->start();
+        });
+
+        // Wait for all commands to finish before carrying on
+        array_walk($commands, function (Process $command) use ($commands) {
+            $exitCode = $command->wait();
+
+            if ($exitCode !== 0) {
+                $this->logger->alert('There was an error in one of the commands -- stopping others.');
+
+                array_walk($commands, function (Process $command) {
+                    $command->stop();
+                });
+            }
+
+            if ($exitCode !== 0) {
+                throw new ServiceException("There was an error exporting the database: " . $command->getErrorOutput());
+            }
+        });
+
+        $this->logger->info('Starting structure and data dump finished.');
+
+        // Once the exports have finished then start the compression.
+        $compressCommand = $this->createProcess(
+            sprintf('echo %s | cat - %s %s | gzip -9 > %s', escapeshellarg($dumpHeader), $structureOutputFile, $dataOutputFile, $compressedFinalFile)
+        );
+
+        $this->logger->info('Starting compress command.');
+
+        $compressCommand->start();
+        $compressCommand->wait();
```

Comment:
We're not checking that the `gzip` succeeded here. It could fail because there's not enough available space, or if the destination file already exists (could force it to overwrite the destination with `-f`).